simulator/infra/macro.h

```diff
@@ -227,4 +227,39 @@ auto test_subtraction_overflow( T_src1 src1, T_src2 src2)
     return std::make_pair( narrow_cast<T>( result), is_overflow);
 }
 
+template<typename T> static
+auto shuffle_stage( T src, uint64_t maskL, uint64_t maskR, int N)
+{
+    auto x = src & ~( maskL | maskR);
+    x |= ( ( src << N) & maskL) | ( ( src >> N) & maskR);
+    return x;
+}
+
+// array with constants for shuffle
+const uint64_t shfl_consts[] = {
+    0x0000000000ff0000ull, 0x000000000000ff00ull,
+    0x000000000f000f00ull, 0x0000000000f000f0ull,
+    0x0000000030303030ull, 0x000000000c0c0c0cull,
+    0x0000000044444444ull, 0x0000000022222222ull,
+    0x0000ffff00000000ull, 0x00000000ffff0000ull,
+    0x00ff000000ff0000ull, 0x0000ff000000ff00ull,
+    0x0f000f000f000f00ull, 0x00f000f000f000f0ull,
+    0x3030303030303030ull, 0x0c0c0c0c0c0c0c0cull,
+    0x4444444444444444ull, 0x2222222222222222ull,
+};
+
+template<typename T, typename U> static
+auto bit_shuffle( T src1, U src2)
```

Comment:
Is there a case when T != U?
Could you please provide an override for uint128 since that code produces compilation failures?